atproto/auth/oauth/oauth.go

```diff
@@ -581,19 +581,42 @@ func (app *ClientApp) StartAuthFlow(ctx context.Context, identifier string) (str
 func (app *ClientApp) ProcessCallback(ctx context.Context, params url.Values) (*ClientSessionData, error) {
 
 	state := params.Get("state")
-	authserverURL := params.Get("iss")
-	authCode := params.Get("code")
-	if state == "" || authserverURL == "" || authCode == "" {
-		return nil, fmt.Errorf("missing required query param")
+	if state == "" {
+		return nil, fmt.Errorf("missing state query param")
 	}
 
 	info, err := app.Store.GetAuthRequestInfo(ctx, state)
 	if err != nil {
 		return nil, fmt.Errorf("loading auth request info: %w", err)
 	}
+	if info.State != state {
+		return nil, fmt.Errorf("callback state doesn't match request info")
+	}
+
+	// NOTE: A corresponding `state` is expected even under error conditions,
+	// hence we check error *after* checking state.
+	errorCode := params.Get("error")
+	if errorCode != "" {
+		var errorUri *syntax.URI
+		parsedUri, err := syntax.ParseURI(params.Get("error_uri"))
+		if err == nil {
+			errorUri = &parsedUri
+		}
+		return nil, &ErrCallback{
+			code:        errorCode,
+			description: params.Get("error_description"),
+			uri:         errorUri,
+		}
+	}
+
+	authserverURL := params.Get("iss")
+	authCode := params.Get("code")
+	if authserverURL == "" || authCode == "" {
```

Comment:
any reason to move these checks down? I would keep the "validating HTTP request" sort of checks up at the top of the method, before doing any database requests.

it would be more friendly to do separate error strings for each missing param as well (not needed for this PR, but could hit it while touching this PR)

Reply:
oh, I see now, an error response won't include any of these. hrm! maybe worth a comment up top that `state` query param is always set, but `error_code`, `iss`, and `code` all depend on type of response?